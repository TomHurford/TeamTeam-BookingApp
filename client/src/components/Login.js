
import React, { Component } from 'react';
import '../styles/Login.css';
const axios = require('axios').default;
const jwtController = require('../utils/jwt.js');


// Create a login component that prints the input email and password to the console
class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "",
      email: "",
      password: "",
      confirmPassword: "",

      //block of state variables relating to showing forms
      disableLoginForm: false,
      disableSignUp: true,
      disableForgotPassword: true,
      disablePopUpMessage: true,
      disableResetPassword: true,
      popUpMessage: "",
    };
  }

  // Update the state of variable associated with a field
  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  };

  // Submit login to backend to receive token for use
  handleSubmitLogin = (event) => {
    if (this.state.username === "" || this.state.password === "") {
      console.log("Empty Fields");
      this.showMessage("You have left a field empty");
      return false;
    }
    event.preventDefault();

    axios
      .post("http://localhost:5001/user/login", {
        email: this.state.email,
        password: this.state.password,
      })
      .then((response) => {
        console.log(response.data.token);
        if (response.data.token) {
          jwtController.setToken(response.data.token);
          window.location = "/profile";
        } else {
          this.showMessage("Error: " + response.data.message);
        }
      })
      .catch((error) => {
        console.log(error.message);
        this.showMessage("Local Client Error: " + error.message);
      });
    return false;
  };

  // Submit signup to backend to receive successful sign up
  handleSubmitSignUp = (event) => {
    if (this.state.confirmPassword === this.state.password) {
      console.log("Passwords Do Not Match");
      this.showMessage("You have left a field empty");
      this.signupToggle();
      return false;
    }

    event.preventDefault();

    axios
      .post("http://localhost:5001/user/signup", {
        name: this.state.name,
        email: this.state.email,
        password: this.state.password,
      })
      .then((response) => {
        if (response.data.success) {
          this.showMessage("Signed Up. Check Email for Verification.");
        } else {
          this.showMessage("Error: " + response.data.message);
        }
      })
      .catch((error) => {
        console.log(error.message);
        this.showMessage("Local Client Error: " + error.message);
      });

    return false;
  };

  // Submit forgot password to backend to receive successful forgot password email
  handleSubmitForgot = (event) => {
    event.preventDefault();

    axios
      .post("http://localhost:5001/user/forgotPassword", {
        email: this.state.email,
      })
      .then((response) => {
        if (response.data.success) {
          this.showMessage("Forgot Password Email Sent If Email Exists.");
        } else {
          this.showMessage("Error: " + response.data.message);
        }
      })
      .catch((error) => {
        console.log(error.message);
        this.showMessage("Local Client Error: " + error.message);
      });

    return false;
  };

  // Submit a new password to be set
  handleSubmitResetPassword = (event) => {
    //     if (this.state.confirmPassword === this.state.password) {
    //         console.log('Passwords Do Not Match');
    //         this.showMessage('You have left a field empty');
    //         signupToggle();
    //         return false;
    //     }
    //     event.preventDefault();
    //     axios.post('http://localhost:5001/user/reset', { verificationCode: password: this.state.password })
    //         .then(response => {
    //             if (response.data.success) {
    //                 this.showMessage('Password Reset');
    //             } else {
    //                 showMessage('Error: ' + response.data.message);
    //             }
    //         })
    //         .catch(error => {
    //             console.log(error.message);
    //             this.showMessage('Local Client Error: ' + error.message);
    //         });
    //     return false;
  };

  // Toggle the signup Form Viewable
  signupToggle = (event) => {
    this.setState({
      disableLoginForm: !this.state.disableLoginForm,
      disableSignUp: !this.state.disableSignUp,
    });
  };

  // Toggle the Forgot Password Form Viewable
  forgotToggle = (event) => {
    this.setState({
      disableLoginForm: !this.state.disableLoginForm,
      disableForgotPassword: !this.state.disableForgotPassword,
    });
  };

  // Toggle the Reset Password Form Viewable
  resetPasswordToggle = (event) => {
    this.setState({
      disableLoginForm: !this.state.disableLoginForm,
      disableResetPassword: !this.state.disableResetPassword,
    });
  };

  // Open the Message Overlay
  showMessage = (message) => {
    this.setState({
      disablePopUpMessage: false,
      popUpMessage: message,
    });
  };

  // Close the Message Overlay
  closeMessage = (event) => {
    this.setState({
      disablePopUpMessage: true,
      disableLoginForm: false,
      disableForgotPassword: true,
      disableSignUp: true,
    });
    return false;
  };

  loginNewUser = (code) => {};

  //Run On Render to check if link holds extra data

  render() {
    return (
      <div>
        <div className="underlay"></div>

        <div
          className="container"
          id="LoginForm"
          disabled={this.state.disableLoginForm}
        >
          <h1>Log In</h1>
          <form onSubmit={this.handleSubmitLogin} action="#">
            <div className="field" style={{ height: "0.9em" }}>
              <label className="buttonLabel" onClick={this.signupToggle}>
                Not A User? Sign Up
              </label>
              <br />
            </div>
            <div className="field">
              <label htmlFor="email">Email</label>
              <br />
              <input
                type="email"
                name="email"
                onChange={this.handleChange}
                required
              />
            </div>
            <div className="field">
              <label htmlFor="password">Password</label>
              <br />
              <input
                type="password"
                name="password"
                minLength="8"
                onChange={this.handleChange}
                required
              />
            </div>
            <div
              className="field"
              style={{
                textAlign: "right",
                height: "0.9em",
                marginTop: 0,
                marginBottom: "10px",
              }}
            >
              <label className="buttonLabel" onClick={this.forgotToggle}>
                Forgot Your Password?
              </label>
              <br />
            </div>
            <button type="submit">Login</button>
          </form>
        </div>

        <div
          className="container"
          id="SignUpForm"
          disabled={this.state.disableSignUp}
        >
          <h1>Sign Up</h1>
          <form onSubmit={this.handleSubmitSignUp} action="#">
            <div className="field" style={{ height: "0.9em" }}>
              <label className="buttonLabel" onClick={this.signupToggle}>
                Already a User? Log In
              </label>
              <br />
            </div>
            <div className="field">
              <label htmlFor="text">Name</label>
              <br />
              <input
                type="text"
                name="name"
                minLength="3"
                onChange={this.handleChange}
                required
              />
            </div>
            <div className="field">
              <label htmlFor="email">Email</label>
              <br />
              <input
                type="email"
                name="email"
                onChange={this.handleChange}
                required
              />
            </div>
            <div className="field">
              <label htmlFor="password">Password</label>
              <br />
              <input
                type="password"
                name="password"
                minLength="8"
                onChange={this.handleChange}
                required
              />
            </div>
            <div className="field">
              <label htmlFor="confirmPassword">Confirm Password</label>
              <br />
              <input
                type="password"
                name="confirmPassword"
                onChange={this.handleChange}
                required
              />
            </div>
            <button type="submit">Sign Up</button>
          </form>
        </div>

        <div
          className="container"
          id="forgotForm"
          disabled={this.state.disableForgotPassword}
        >
          <h1>Forgot Password</h1>
          <form onSubmit={this.handleSubmitForgot} action="#">
            <div className="field">
              <label htmlFor="email">Email</label>
              <br />
              <input
                type="email"
                name="email"
                onChange={this.handleChange}
                required
              />
            </div>
            <div
              className="field"
              style={{
                textAlign: "right",
                height: "0.9em",
                marginTop: 0,
                marginBottom: "10px",
              }}
            >
              <label className="buttonLabel" onClick={this.forgotToggle}>
                Remember Your Password?
              </label>
              <br />
            </div>
            <button type="submit">Send Reset Password Email</button>
          </form>
        </div>

        <div
          className="container"
          id="resetPasswordForm"
          disabled={this.state.disableResetPassword}
        >
          <h1>Reset Password</h1>
          <form onSubmit={this.handleSubmitResetPassword} action="#">
            <div className="field">
              <label htmlFor="password">Password</label>
              <br />
              <input
                type="password"
                name="password"
                minLength="8"
                onChange={this.handleChange}
                required
              />
            </div>
            <div className="field">
              <label htmlFor="confirmPassword">Confirm Password</label>
              <br />
              <input
                type="password"
                name="confirmPassword"
                onChange={this.handleChange}
                required
              />
            </div>
            <button type="submit">Reset Password</button>
          </form>
        </div>

        <div className="overlay" disabled={this.state.disablePopUpMessage}>
          <div className="container" id="popupForm">
            <form onSubmit={this.closeMessage} action="#">
              <div className="field">
                <label style={{ textAlign: "center" }}>
                  {this.state.popUpMessage}
                </label>
                <br />
              </div>
              <button type="submit">Close</button>
            </form>
          </div>
        </div>
      </div>
    );
  }
}

export default Login;
