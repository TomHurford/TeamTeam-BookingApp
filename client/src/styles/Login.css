

.container {
    width: 30vw;
    min-width: 200px;
    left: 35vw;
    position: relative;
    aspect-ratio: 0.75;
    position: relative;
    align-self: center;
    text-align: center;
    display: block;
}
.container[disabled] {
    height: 0;
    width: 0;
    overflow: hidden;
    display: none;
}

@media (max-width: 667px) {
    .container {
        left: calc( 50vw - 100px );
    }
}

.container > h1 {
    font-size: 3em;
}

.container > form {
    width: 100%;
    margin-bottom: 20px;
}

.container  .field {
    margin-top: 10px;
    width: 100%;
    aspect-ratio: 4;
    text-align: left;
    font-size: 1.1em;
}
.container  .field > label {
    width: 100%;
    padding: 5px;
}
.container  .field > label.buttonLabel {
    width: 100%;
    padding: 5px;
    font-size: 0.7em;
    line-height: 0.9em;
}
.container  .field > label.buttonLabel:hover {
    opacity: 0.8;
    cursor: pointer;
}
.container .field > input {
    padding-left: 15px;
    padding-right: 15px;
    width: calc( 100% - 30px );
    aspect-ratio: 10;
    box-shadow: 1px 1px 3px rgba(135, 135, 135, 0.452);
    border-radius: 10px;
    border: 1px;
    outline: 5px;
    font-size: 1.1em;
}
.container  .field > label.errortext {
    color: red;
    font-size: 0.7em;
    line-height: 0.9em;
    padding-top: 0;
}
.container  .field > input:active {
    border: 2px;
}
.container .field > textarea {
    padding-left: 15px;
    padding-right: 15px;
    width: calc( 100% - 30px );
    box-shadow: 1px 1px 3px rgba(135, 135, 135, 0.452);
    border-radius: 10px;
    border: 1px;
    outline: 5px;
    font-size: 1.1em;
}
.container  .field > textarea:active {
    border: 2px;
}

.container  button {
    padding-left: 15px;
    padding-right: 15px;
    width: 100%;
    aspect-ratio: 10;
    box-shadow: 1px 1px 3px rgba(135, 135, 135, 0.452);
    border-radius: 10px;
    border: 1px;
    outline: 5px;
    font-size: 1.1em;
    text-align: center;
    margin: auto;
    margin-top: 10px;
}
.container  button:hover {
    transform: scale(1.1);
    transition: 0.5s;
    cursor: pointer;
}
.container  button:active {
    transform: scale(1.05);
    transition: 0.5s;
    opacity: 0.8;
}

.underlay {
    width: 100vw;
    height: 100vh;
    position: fixed;
    top: 0;
    left: 0;
    z-index: -10000;
    /* Unsplash Royalty free License */
    background: url('https://images.unsplash.com/photo-1567201864585-6baec9110dac?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=687&q=80');
    background-repeat: no-repeat;
    background-size: cover;
    background-position: center;
    opacity: 0.5;
}