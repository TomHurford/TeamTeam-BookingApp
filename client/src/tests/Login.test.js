import React from 'react';
import renderer from 'react-test-renderer';
import Login from '../components/Login';
// Snapshot test for Login page
test('Login snapshot test', () => {
  const component = renderer.create(<Login />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});