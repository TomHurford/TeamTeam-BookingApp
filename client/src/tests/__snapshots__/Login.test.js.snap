// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login snapshot test 1`] = `
<div
  className="page-container"
>
  <div
    className="underlay"
  />
  <div
    className="container"
    disabled={false}
    id="LoginForm"
  >
    <h1>
      Log In
    </h1>
    <div
      className="field"
      style={
        Object {
          "height": "0.9em",
        }
      }
    >
      <label
        className="buttonLabel"
        onClick={[Function]}
      >
        Not A User? Sign Up
      </label>
      <br />
    </div>
    <div
      className="field"
    >
      <label
        htmlFor="email"
      >
        Email
      </label>
      <br />
      <input
        name="loginEmail"
        onChange={[Function]}
        required={true}
        type="email"
      />
    </div>
    <div
      className="field"
    >
      <label
        htmlFor="password"
      >
        Password
      </label>
      <br />
      <input
        minLength="8"
        name="loginPassword"
        onChange={[Function]}
        required={true}
        type="password"
      />
    </div>
    <div
      className="field"
      style={
        Object {
          "height": "0.9em",
          "marginBottom": "10px",
          "marginTop": 0,
          "textAlign": "right",
        }
      }
    >
      <label
        className="buttonLabel"
        onClick={[Function]}
      >
        Forgot Your Password?
      </label>
      <br />
    </div>
    <button
      className="button"
      name="loginbutton"
      onClick={[Function]}
    >
      Login
    </button>
  </div>
  <div
    className="container"
    disabled={true}
    id="SignUpForm"
  >
    <h1>
      Sign Up
    </h1>
    <div
      className="field"
      style={
        Object {
          "height": "0.9em",
        }
      }
    >
      <label
        className="buttonLabel"
        onClick={[Function]}
      >
        Already a User? Log In
      </label>
      <br />
    </div>
    <div
      className="field"
    >
      <label
        htmlFor="text"
      >
        Name
      </label>
      <br />
      <input
        minLength="3"
        name="signupName"
        onChange={[Function]}
        required={true}
        type="text"
      />
    </div>
    <div
      className="field"
    >
      <label
        htmlFor="email"
      >
        Email
      </label>
      <br />
      <input
        name="signupEmail"
        onChange={[Function]}
        required={true}
        type="email"
      />
    </div>
    <div
      className="field"
    >
      <label
        htmlFor="password"
      >
        Password
      </label>
      <br />
      <input
        minLength="8"
        name="signupPassword"
        onChange={[Function]}
        required={true}
        type="password"
      />
    </div>
    <div
      className="field"
    >
      <label
        htmlFor="confirmPassword"
      >
        Confirm Password
      </label>
      <br />
      <input
        name="signupConfirmPassword"
        onChange={[Function]}
        required={true}
        type="password"
      />
    </div>
    <button
      className="button"
      name="signup-button"
      onClick={[Function]}
    >
      Sign Up
    </button>
  </div>
  <div
    className="container"
    disabled={true}
    id="forgotForm"
  >
    <h1>
      Forgot Password
    </h1>
    <div
      className="field"
    >
      <label
        htmlFor="email"
      >
        Email
      </label>
      <br />
      <input
        name="forgotEmail"
        onChange={[Function]}
        required={true}
        type="email"
      />
    </div>
    <div
      className="field"
      style={
        Object {
          "height": "0.9em",
          "marginBottom": "10px",
          "marginTop": 0,
          "textAlign": "right",
        }
      }
    >
      <label
        className="buttonLabel"
        onClick={[Function]}
      >
        Remember Your Password?
      </label>
      <br />
    </div>
    <button
      className="button"
      name="forgot-button"
      onClick={[Function]}
    >
      Send Reset Password Email
    </button>
  </div>
  <div
    className="container"
    disabled={true}
    id="resetPasswordForm"
  >
    <h1>
      Reset Password
    </h1>
    <div
      className="field"
    >
      <label
        htmlFor="password"
      >
        Password
      </label>
      <br />
      <input
        minLength="8"
        name="resetPassword"
        onChange={[Function]}
        required={true}
        type="password"
      />
    </div>
    <div
      className="field"
    >
      <label
        htmlFor="confirmPassword"
      >
        Confirm Password
      </label>
      <br />
      <input
        name="resetConfirmPassword"
        onChange={[Function]}
        required={true}
        type="password"
      />
    </div>
    <button
      data-testid="reset"
      onClick={[Function]}
      type="submit"
    >
      Reset Password
    </button>
  </div>
  <div
    className="overlay"
    disabled={true}
  >
    <div
      className="container"
      id="popupForm"
    >
      <div
        className="field"
      >
        <label
          style={
            Object {
              "textAlign": "center",
            }
          }
        />
        <br />
      </div>
      <button
        onClick={[Function]}
        type="submit"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;
